File: exercise/ts/day21/src/tour/tourCalculator.ts
import * as E from 'fp-ts/Either';

export type Step = {
    time: string;
    label: string;
    deliveryTime: number;
};

export class TourCalculator {
    private steps: Step[];
    private calculated: boolean = false;
    private _deliveryTime: number = 0;

    constructor(steps: Step[]) {
        this.steps = steps;
    }

    get deliveryTime(): number {
        return this._deliveryTime;
    }

    calculate(): E.Either<string, string> {
        if (!this.steps || this.steps.length === 0) {
            return E.left('No locations !!!');
        } else {
            const result: string[] = [];

            for (const s of this.steps
                .sort((a, b) => (a.time > b.time ? 1 : -1))) {
                if (!this.calculated) {
                    this._deliveryTime += s.deliveryTime;
                    result.push(this.fLine(s, this._deliveryTime));
                }
            }

            const str = new Date(this._deliveryTime * 1000).toISOString().substring(11);
            result.push(`Delivery time | ${str}`);
            this.calculated = true;

            return E.right(result.join('\n'));
        }
    }

    private fLine(step: Step, x: number): string {
        if (!step) throw new Error('Invalid operation');
        return `${step.time} : ${step.label} | ${step.deliveryTime} sec`;
    }
}

File: exercise/C#/day21/Tour/TourCalculator.cs
using System.Text;
using LanguageExt;
using static LanguageExt.Prelude;

namespace Tour
{
    public record Step(TimeOnly Time, string Label, int DeliveryTime);

    public class TourCalculator
    {
        public List<Step> Steps { get; set; }

        public bool Calculated
        {
            get { return calculated; }
        }

        public double DeliveryTime
        {
            get { return _deliveryTime; }
        }

        private double _deliveryTime = 0;
        private bool calculated = false;

        public TourCalculator(List<Step> steps)
        {
            Steps = steps;
        }

        public Either<string, string> Calculate()
        {
            if (Steps.IsNull() || Steps.Count == 0)
            {
                return Left("No locations !!!");
            }
            else
            {
                var result = new StringBuilder();

                foreach (var s in Steps.OrderBy(x => x.Time))
                {
                    if (!calculated)
                    {
                        this._deliveryTime += s.DeliveryTime;
                        result.AppendLine(fLine(s, _deliveryTime));
                    }
                }

                string hhMmSs = @"hh\:mm\:ss";
                string str = TimeSpan.FromSeconds(this._deliveryTime).ToString(hhMmSs);
                result.AppendLine($"Delivery time | {str}");
                calculated = true;

                return Right(result.ToString());
            }
        }

        private string fLine(Step step, double x)
        {
            if (step == null)
                throw new InvalidOperationException();
            else
                return $"{step.Time} : {step.Label} | {step.DeliveryTime} sec";
        }
    }
}